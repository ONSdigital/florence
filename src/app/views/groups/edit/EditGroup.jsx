import React, { useEffect, useState, useCallback } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router";
import { withRouter } from "react-router";
import isEqual from "lodash/isEqual";
import isEmpty from "lodash/isEmpty";
import BackButton from "../../../components/back-button";
import Loader from "../../../components/loader";
import Input from "../../../components/Input";
import validate from "./validate";
import Members from "../members";
import DeletePanel from "../delete";
import FormFooter from "../../../components/form-footer/FormFooter";
import UsersTable from "../../../components/table";
import Magnifier from "../../../icons/Magnifier";
import clsx from "clsx";

const EditGroup = props => {
    const id = props.params.id;

    useEffect(() => {
        if (id) {
            props.loadGroup(id);
        }
    }, []);

    useEffect(() => {
        if (id) {
            props.loadMembers(id);
        }
    }, []);

    useEffect(() => {
        props.loadUsers();
    }, []);

    const { group, loading, users, updateGroup, loadingUsers, members, loadingMembers, updateGroupMembers } = props;
    const [values, setValues] = useState(null);
    const [search, setSearch] = useState("");
    const [errors, setErrors] = useState({});
    const [isSubmitting, setIsSubmitting] = useState(false);

    const [groupMembers, setGroupMembers] = useState(members);
    const [availableUsers, setAvailableUsers] = useState(users); // all available - those in groups groupMembers

    const hasErrors = !isEmpty(errors);
    const hasNewValues = !isEqual(values, group);
    const hasUpdatedMembers = !isEqual(groupMembers, members);
    const specialGroup = group?.precedence === (1 || 2);

    const routerWillLeave = nextLocation => {
        if ((hasNewValues || hasUpdatedMembers) && !isSubmitting) return "Your work is not saved! Are you sure you want to leave?";
    };

    useEffect(() => {
        setAvailableUsers(users);
        setGroupMembers(members);
    }, [users, members]);

    useEffect(() => {
        const newAvailableUsers = availableUsers.filter(user => !groupMembers.some(m => m.id === user.id));
        setAvailableUsers(newAvailableUsers);
    }, [groupMembers]);

    useEffect(() => {
        props.router.setRouteLeaveHook(props.route, routerWillLeave);
    });

    useEffect(() => {
        if (group) {
            setValues({ ...group }); // ths is not necessary but in future we will edit more fields
        }
    }, [group]);

    useEffect(() => {
        const str = search.toLowerCase();
        const newAvailableUsers = users.filter(user => !groupMembers.some(m => m.id === user.id));
        const result = newAvailableUsers.filter(user => user.name?.toLowerCase().includes(str) || users.email?.toLowerCase().includes(str));

        setAvailableUsers(result);
    }, [search, groupMembers, users]);

    const handleSubmit = e => {
        if (e) e.preventDefault();
        setErrors(validate(values));
        setIsSubmitting(true);

        if (hasErrors || values.name === "") return;

        updateGroup(id, { name: values.name, precedence: 10 });
        if (hasUpdatedMembers) {
            updateGroupMembers(
                id,
                groupMembers.map(({ id }) => ({ user_id: id }))
            );
        }
    };

    const handleInputChange = useCallback(e => {
        const key = e.target.name;
        const val = e.target.value;
        setIsSubmitting(false);

        if (hasErrors) {
            delete errors[e.target.name];
            setErrors(errors);
        }
        setValues(values => ({ ...values, [key]: val }));
    }, []);

    const handleSearchChange = e => setSearch(e.target.value);

    const handleRemove = user => {
        setGroupMembers(prevState => prevState.filter(member => member.id !== user.id));
        setAvailableUsers(prevState => prevState.concat(user));
    };

    const handleAdd = member => {
        setGroupMembers(prevState => prevState.concat(member));
        setAvailableUsers(prevState => prevState.filter(user => user.id !== member.id));
    };

    if (loading) return <Loader classNames="grid grid--align-center grid--align-self-center grid--full-height" />;
    if (!group) return <h1>No group found.</h1>;

    return (
        <div className="form form-with-sticky-footer">
            <div className="grid grid--justify-space-around">
                <div className="grid__col-11 grid__col-md-9">
                    <BackButton classNames="margin-top--2" />
                    <h1 className="margin-top--1 margin-bottom--1">{group.name}</h1>
                    <div className="grid grid--justify-space-between">
                        <div className="grid__col-md-6 margin-top--half">
                            {!specialGroup && (
                                <Input
                                    error={errors?.name}
                                    id="name"
                                    label="Name"
                                    name="name"
                                    type="text"
                                    value={values ? values.name : ""}
                                    onChange={handleInputChange}
                                />
                            )}
                            {loadingMembers && <Loader classNames="grid grid--align-center grid--align-self-center grid--full-height" />}
                            {groupMembers && <Members members={groupMembers} handleRemove={handleRemove} />}
                            {!specialGroup && <DeletePanel id={id} />}
                        </div>
                        <div className="grid__col-md-5">
                            <h2 className="font-size--16">Add member to team</h2>
                            {loadingUsers ? (
                                <Loader classNames="grid grid--align-center grid--align-self-center" />
                            ) : (
                                <>
                                    <div className="search__input-group margin-bottom--1">
                                        <Magnifier classes="search__icon-magnifier" viewBox="0 0 28 28" />
                                        <label htmlFor="search" className="visually-hidden">
                                            Search users by name or email
                                        </label>
                                        <input
                                            role="search"
                                            name="search"
                                            placeholder="Search users by name"
                                            value={search ? search : ""}
                                            onChange={handleSearchChange}
                                        />
                                    </div>
                                    <UsersTable testid="users-table" items={availableUsers} handleClick={handleAdd} />
                                </>
                            )}
                        </div>
                    </div>
                </div>
                <FormFooter hasNewValues={hasNewValues} hasErrors={hasErrors} loading={loading} handleSubmit={handleSubmit} />
            </div>
        </div>
    );
};

EditGroup.prototype = {
    params: PropTypes.objectOf({ id: PropTypes.string.isRequired }),
    group: PropTypes.exact({
        created: PropTypes.string,
        id: PropTypes.string,
        members: PropTypes.array,
        name: PropTypes.string,
        precedence: PropTypes.number,
    }).isRequired,
    updateGroup: PropTypes.func.isRequired,
    loadGroup: PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired,
};

export default withRouter(EditGroup);
