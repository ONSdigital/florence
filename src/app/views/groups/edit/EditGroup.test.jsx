import React from "react";
import renderer from "react-test-renderer";
// import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "../../../utilities/tests/test-utils";
import { group, specialGroup } from "../../../utilities/tests/mockData";
import EditGroup from "./EditGroup";

const users = [
    {
        active: true,
        email: "foo@ons.gov.uk",
        forename: "Test",
        groups: [],
        id: "0234",
        lastname: "Foo",
        status: "CONFIRMED",
        status_notes: "",
    },
    {
        active: true,
        email: "test.user@ons.gov.uk",
        forename: "test",
        groups: [],
        id: "1234",
        lastname: "user",
        status: "CONFIRMED",
        status_notes: "",
    },
    {
        active: true,
        email: "john.doe@ons.gov.uk",
        forename: "John",
        groups: [],
        id: "0456",
        lastname: "Doe",
        status: "CONFIRMED",
        status_notes: "",
    },
];
const props = {
    group: group,
    loading: false,
    params: { id: "0", router: setRouteLeaveHook },
    rootPath: "test",
    router: { setRouteLeaveHook: jest.fn() },
    loadingUsers: false,
    members: [
        {
            active: true,
            email: "test.user-199@ons.gov.uk",
            forename: "test",
            groups: [],
            id: "test-user-199",
            lastname: "user-199",
            status: "CONFIRMED",
            status_notes: "",
        },
    ],
    loadingMembers: false,
    users: users,
    loadGroup: jest.fn(),
    loadingMembers: jest.fn(),
    loadMembers: jest.fn(),
    loadUsers: jest.fn(),
    updateGroup: jest.fn(),
    updateGroupMembers: jest.fn(),
};

const setRouteLeaveHook = jest.fn();
jest.mock("../delete", () => "div"); // faking connected component

describe("EditGroup", () => {
    it("matches the snapshot", () => {
        const tree = renderer.create(<EditGroup.WrappedComponent {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("requests group on load by params", () => {
        render(<EditGroup.WrappedComponent {...props} />);
        expect(props.loadGroup).toHaveBeenCalledWith("0");
    });

    it("shows Back Button", () => {
        render(<EditGroup.WrappedComponent {...props} />);
        expect(screen.getByRole("link", { name: "Back" })).toBeInTheDocument();
    });

    it("shows message if no group has been found", () => {
        render(<EditGroup.WrappedComponent {...props} group={null} />);
        expect(screen.getByText("No group found.")).toBeInTheDocument();
    });

    it("shows loader when fetching group", () => {
        render(<EditGroup.WrappedComponent {...props} group={null} loading={true} />);
        expect(screen.getByTestId("loader")).toBeInTheDocument();
    });

    it("shows group details and name input to edit", () => {
        render(<EditGroup.WrappedComponent {...props} />);
        expect(screen.getByRole("heading", { level: 1, name: "Boo is fine" })).toBeInTheDocument();
        expect(screen.getByLabelText("Name")).toHaveValue("Boo is fine");
        expect(screen.getByRole("button", { name: "Save changes" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
        expect(screen.queryByText(/You have unsaved changes/i)).not.toBeInTheDocument();
    });

    describe("when editing normal group name", () => {
        it("validates for emptiness", () => {
            render(<EditGroup.WrappedComponent {...props} />);

            userEvent.clear(screen.getByLabelText(/name/i));

            expect(screen.getByLabelText("Name")).toHaveValue("");

            userEvent.click(screen.getByRole("button", { name: "Save changes" }));

            expect(screen.getByText(/Please enter a name/i)).toBeInTheDocument();
            expect(screen.getByRole("button", { name: "Save changes" })).toBeDisabled();
        });

        it("Updates group name and members", () => {
            render(<EditGroup.WrappedComponent {...props} />);

            userEvent.clear(screen.getByLabelText(/name/i));
            userEvent.paste(screen.getByLabelText(/name/i), "Foo");

            expect(screen.getByText(/You have unsaved changes/i)).toBeInTheDocument();
            expect(screen.getByLabelText("Name")).toHaveValue("Foo");

            userEvent.click(screen.getByRole("button", { name: "Save changes" }));

            expect(props.updateGroup).toHaveBeenCalledWith("0", { name: "Foo", precedence: 10 });
        });
    });
});
