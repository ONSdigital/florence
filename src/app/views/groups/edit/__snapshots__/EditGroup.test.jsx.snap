// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditGroup matches the snapshot 1`] = `
<div
  className="form form-with-sticky-footer"
>
  <div
    className="grid grid--justify-space-around"
  >
    <div
      className="grid__col-11 grid__col-md-9"
    >
      <div
        className="margin-top--2"
      >
        ◀ 
        <a
          onClick={[Function]}
          role="link"
          style={Object {}}
        >
          Back
        </a>
      </div>
      <h1
        className="margin-top--1 margin-bottom--1"
      >
        Boo is fine
      </h1>
      <div
        className="grid grid--justify-space-between"
      >
        <div
          className="grid__col-md-6 margin-top--half"
        >
          <div
            className="form__input"
          >
            <label
              className="form__label"
              htmlFor="name"
            >
              Name
            </label>
            <input
              autoComplete="new-password"
              autoFocus={false}
              className="input"
              disabled={false}
              id="name"
              name="name"
              onChange={[Function]}
              onFocus={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="grid grid--align-center grid--align-self-center grid--full-height"
            data-testid="loader"
          >
            <div
              className="loader loader--large loader--dark"
            />
          </div>
          <div
            className="simple-table__scroll margin-bottom--1"
            data-testid="members"
          >
            <h2>
              Members
            </h2>
            <div
              className="chip__container chip__container--gap-10 "
            >
              <span
                className="chip"
              >
                <svg
                  aria-label="Person icon"
                  className="chip__icon"
                  fill="none"
                  role="img"
                  viewBox="0 0 15 15"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.667 7.5a3.332 3.332 0 1 0 0-6.667 3.332 3.332 0 1 0 0 6.667Zm0 1.667C4.442 9.167 0 10.283 0 12.5v1.667h13.333V12.5c0-2.217-4.441-3.333-6.666-3.333Z"
                    fill="#707071"
                  />
                </svg>
                <span
                  className="chip__text"
                >
                  test user-199 (test.user-199@ons.gov.uk)
                </span>
                <span
                  className="chip__remove-container"
                >
                  <button
                    aria-label="remove"
                    className="chip__remove"
                    onClick={[Function]}
                  />
                </span>
              </span>
            </div>
          </div>
          <div
            id="0"
          />
        </div>
        <div
          className="grid__col-md-5"
        >
          <h2
            className="font-size--16"
          >
            Add member to team
          </h2>
          <div
            className="search__input-group margin-bottom--1"
          >
            <svg
              className="search__icon-magnifier"
              role="presentation"
              viewBox="0 0 28 28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m20.011 16.352h-1.2647l-0.4482-0.4014c1.5689-1.6947 2.5134-3.8948 2.5134-6.2881 0-5.3368-4.6587-9.6627-10.406-9.6627-5.7472 0-10.406 4.3259-10.406 9.6627 0 5.3367 4.6587 9.6626 10.406 9.6626 2.5775 0 4.9469-0.877 6.7719-2.3339l0.4323 0.4163v1.1743l8.0045 7.418 2.3854-2.215-7.9886-7.4328zm-9.6055 0c-3.9862 0-7.2041-2.988-7.2041-6.6895s3.2178-6.6895 7.2041-6.6895c3.9863 0 7.2042 2.988 7.2042 6.6895s-3.2179 6.6895-7.2042 6.6895z"
                fill="#808080"
              />
            </svg>
            <label
              className="visually-hidden"
              htmlFor="search"
            >
              Search users by name or email
            </label>
            <input
              name="search"
              onChange={[Function]}
              placeholder="Search users by name"
              role="search"
              value=""
            />
          </div>
          <table
            className="simple-table simple-table__scroll"
            data-testid="users-table"
            role="table"
          >
            <tbody>
              <tr
                className="simple-table__row grid grid--align-center"
              >
                <td
                  className="grid__col-10"
                />
                <td
                  className="grid__col-2"
                >
                  <button
                    className="btn btn--info"
                    onClick={[Function]}
                  >
                    Add
                  </button>
                </td>
              </tr>
              <tr
                className="simple-table__row grid grid--align-center"
              >
                <td
                  className="grid__col-10"
                />
                <td
                  className="grid__col-2"
                >
                  <button
                    className="btn btn--info"
                    onClick={[Function]}
                  >
                    Add
                  </button>
                </td>
              </tr>
              <tr
                className="simple-table__row grid grid--align-center"
              >
                <td
                  className="grid__col-10"
                />
                <td
                  className="grid__col-2"
                >
                  <button
                    className="btn btn--info"
                    onClick={[Function]}
                  >
                    Add
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <div
      className="grid grid--justify-space-around padding-bottom--1 padding-top--1 form-footer"
      data-testid="form-footer"
    >
      <div
        className="grid__col-9"
      >
        <div
          className="grid grid--align-baseline"
        >
          <button
            className="btn btn--positive btn--margin-right"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            Save changes
          </button>
          <a
            className="btn btn--invert-primary btn--margin-right"
            onClick={[Function]}
            role="button"
            style={Object {}}
          >
            Cancel
          </a>
          <span>
            <svg
              aria-label="warning icon"
              className="svg-icon--action-bar"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256,34.297L0,477.703h512L256,34.297z M256,422.05c-9.22,0-16.696-7.475-16.696-16.696s7.475-16.696,16.696-16.696 c9.22,0,16.696,7.475,16.696,16.696S265.22,422.05,256,422.05z M239.304,344.137V177.181h33.391v166.956H239.304z"
              />
            </svg>
             You have unsaved changes
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
